mio/src/lib.rs

```diff
@@ -1,111 +1,145 @@
 extern crate mio;
 extern crate futures;
+extern crate fnv;
 
+use std::hash::BuildHasherDefault;
 use std::collections::HashMap;
 use std::io::{self, ErrorKind, Read, Write};
 use std::net::SocketAddr;
 use std::panic;
 use std::slice;
 use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::mpsc::{channel, TryRecvError};
 
-use futures::{Future, promise, Complete, PollError};
+use fnv::FnvHasher;
 
-pub type IoFuture<T> = Future<Item=T, Error=io::Error>;
+use futures::{Future, Tokens, promise, Complete, Wake, PollError, PollResult};
 
-pub struct Loop {
-    io: mio::Poll,
-    tx: mio::channel::Sender<Message>,
-    rx: mio::channel::Receiver<Message>,
-    next: usize,
-    done: HashMap<usize, Complete<(), io::Error>>,
+pub struct MioEvent {
+    events: mio::EventSet,
+    source: Source,
+    loop_handle: LoopHandle,
 }
 
-enum Message {
-    Wait(Complete<(), io::Error>, mio::EventSet, Arc<mio::Evented + Send + Sync>),
-    Register(Arc<mio::Evented + Send + Sync>),
+impl MioEvent {
+    fn into_future<C: PollCompletion>(self, completion: C) -> MioFuture<C> {
+        MioFuture {
+            event: self,
+            completion: completion,
+        }
+    }
+}
+
+pub trait PollCompletion {
+    type Item: Send + 'static;
+    type Error: Send + 'static;
+
+    fn poll_completion(&mut self) -> Option<PollResult<Self::Item, Self::Error>>;
 }
 
+pub struct MioFuture<C> {
+    event: MioEvent,
+    completion: C,
+}
+
+impl<C> Future for MioFuture<C> where C: PollCompletion + Send + 'static {
+    type Item = C::Item;
+    type Error = C::Error;
+
+    fn poll(&mut self, tokens: &Tokens) -> Option<PollResult<C::Item, C::Error>> {
+        // TODO: filter by tokens
+        self.completion.poll_completion()
+    }
+
+    fn schedule(&mut self, wake: Arc<Wake>) {
+        // TODO: record token
+        self.event.loop_handle.schedule(Interest {
+            waiter: wake,
+            source: self.event.source.clone(),
+            events: self.event.events,
+            first_time: false, // assume already registered
+        });
+    }
+
+    fn tailcall(&mut self) -> Option<Box<Future<Item=Self::Item, Error=Self::Error>>> {
+        None
+    }
+}
+
+
+#[derive(Clone)]
 pub struct TcpListener {
     tcp: Arc<mio::tcp::TcpListener>,
-    tx: mio::channel::Sender<Message>,
+    loop_handle: LoopHandle,
 }
 
-pub struct Error<T> {
-    err: io::Error,
-    data: T,
+pub struct TcpListenerAccept {
+    inner: TcpListener
 }
 
 impl TcpListener {
     pub fn local_addr(&self) -> io::Result<SocketAddr> {
         self.tcp.local_addr()
     }
 
-    pub fn accept(&self) -> Box<IoFuture<(TcpStream, SocketAddr)>> {
-        match self.tcp.accept() {
-            Err(e) => return futures::failed(e).boxed(),
+    pub fn accept(&self) -> TcpListenerAccept {
+        TcpListenerAccept { inner: self.clone() }
```

Comment:
No -- I think you only want that split in the case that there's more than one `Future` interface you want to provide. So for `TcpStream`, exposing `MioEvent`s means that we can easily build up a lot of different `Future` abstractions over the core event, e.g. for different buffering strategies. But that seems much less useful for listeners -- the existing `Future` interface (or, ultimately, a stream interface) should be enough to build anything else you might want.